### PR TITLE
Enable SocketsHttpHandler cancellation support

### PR DESCRIPTION
**Pull request by stephentoub**
This change significantly improves the cancellation support in SocketsHttpHandler.  Previously we were passing the CancellationToken around to every method, eventually bottoming out in calls to the underlying Stream which then ends up passing them down to the underlying Socket.  But today Socket's support for cancellation is minimal, only doing up-front checks; if cancellation is requested during the socket operation rather than before, the request will be ignored.  Since HttpClient implements features like timeouts on top of cancellation support, it's important to do better than this.

The change implements cancellation by registering with the CancellationToken to dispose of the connection.  This will cause any reads/writes to wake up.  We then translate resulting exceptions into cancellation exceptions.  When in the main SendAsync method, we register once for the whole body of the operation until the point that we're returning the response message.  For individual operations on the response content stream, we register per operation; however, when feasible we try to avoid the registration costs by only registering if operations don't complete synchronously.  We also account for the case that on Unix, closing the connection may result in read operations waking up not with an exception but rather with EOF, which we also need to translate into cancellation when appropriate.

Along the way I cleaned up a few minor issues as well.

I also added a bunch of cancellation-related tests:
- Test cancellation occurring while sending request content
- Test cancellation occurring while receiving response headers
- Test cancellation occurring while receiving response body and using a buffered operation
- Test that all of the above are triggerable with CancellationTokenSource.Cancel, HttpClient.CancelPendingRequests, and HttpClient.Dispose
- Test cancellation occurring while receiving response body and using an unbuffered operation, either a ReadAsync or CopyToAsync on the response stream
- Test that a CancelPendingRequests doesn't affect unbuffered operations on the response stream

There are deficiencies here in the existing handlers, and tests have been selectively disabled accordingly.  SocketsHttpHandler passes now for all of them.  I will open issues for the non-netfx issues and include the corresponding numbers in comments before merging.

cc: @geoffkizer, @Priya91, @wfurt, @davidsh 
Fixes https://github.com/dotnet/corefx/issues/23131
Fixes https://github.com/dotnet/corefx/issues/23141

### CODE REVIEW

File: src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs [hunk 1]
```diff
@@ -2,161 +2,442 @@
...
     {
-        private readonly ITestOutputHelper _output;
+        [OuterLoop]
+        [Theory]
+        [MemberData(nameof(TwoBoolsAndCancellationMode))]
+        public async Task PostAsync_CancelDuringRequestContentSend_TaskCanceledQuickly(bool chunkedTransfer, bool connectionClose, CancellationMode mode)
+        {
+            if (IsWinHttpHandler || IsNetfxHandler)
+            {
+                // hangs / doesn't cancel
```
Comment (davidsh) [inline src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs]: Is there a new issue number for the hang for WinHttpHandler?

**Comment by stephentoub** [comment]
@dotnet-bot test NETFX x86 Release Build please (https://github.com/dotnet/corefx/issues/26089)

**Comment by stephentoub** [comment]
@dotnet-bot test OSX x64 Debug Build please (https://github.com/dotnet/corefx/issues/27038)

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop NETFX x86 Release Build please
@dotnet-bot test Outerloop NETFX x86 Release Build please

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please
@dotnet-bot test Outerloop Linux x64 Release Build please

File: src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs]: Not currently.  Per "I will open issues for the non-netfx issues and include the corresponding numbers in comments before merging", I'll be opening an issue and adding the issue number before submitting.  I took a quick look at the source to see if it was obvious how to address it, but it wasn't immediately clear to me.

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Linux x64 Release Build please

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs [hunk 2]
```diff
@@ -84,59 +84,77 @@ public ChunkedEncodingReadStream(HttpConnection connection) : base(connection)
...
+                cancellationToken.ThrowIfCancellationRequested();
+
                 if (_connection == null || destination.Length == 0)
                 {
                     // Response body fully consumed or the caller didn't ask for any data
                     return 0;
                 }
 
-                if (_chunkBytesRemaining == 0)
+                CancellationTokenRegistration ctr = _connection.RegisterCancellation(cancellationToken);
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs]: Can we defer this to after the _chunkBytesRemaining == 0 test?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs [hunk 3]
```diff
@@ -145,23 +163,41 @@ public override Task<int> ReadAsync(byte[] buffer, int offset, int count, Cancel
...
+                cancellationToken.ThrowIfCancellationRequested();
 
                 if (_connection == null)
                 {
                     // Response body fully consumed
                     return;
                 }
 
-                if (_chunkBytesRemaining > 0)
+                CancellationTokenRegistration ctr = _connection.RegisterCancellation(cancellationToken);
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs]: Same

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingWriteStream.cs [hunk 4]
```diff
@@ -13,24 +13,23 @@ private sealed class ChunkedEncodingWriteStream : HttpContentWriteStream
...
             private static readonly byte[] s_finalChunkBytes = { (byte)'0', (byte)'\r', (byte)'\n', (byte)'\r', (byte)'\n' };
 
-            public ChunkedEncodingWriteStream(HttpConnection connection, CancellationToken cancellationToken) :
-                base(connection, cancellationToken)
+            public ChunkedEncodingWriteStream(HttpConnection connection) : base(connection)
             {
             }
 
-            public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken ignored)
+            public override Task WriteAsync(byte[] buffer, int offset, int count, CancellationToken ignored) // token ignored as it comes from SendAsync
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingWriteStream.cs]: Should we assert that it is None?  Same below

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs (hunk 2 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs]: Will that buys us much?  If chunkBytesRemaining == 0 then we're going to do TryGetNextChunkAsync, for which we want the registration in place as it potentially does a read, and if chunkBytesRemaining != 0, then we're going to do a ReadAsync, for which we want the registration in place.  Am I misunderstanding?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingWriteStream.cs (hunk 4 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingWriteStream.cs]: Yup, we can do that.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs (hunk 2 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs]: Yes, you're right -- I missed that we need the registration for the subsequent ReadAsync.  

That said, in the scenario where we have buffered bytes that we can use to satisfy the read, we will unnecessarily register for the cancellation token.  Not sure this matters in practice, and it's not obvious how to fix this in the code, given how HttpConnection.ReadAsync works with buffering and cancellation.  Still, might be worth a comment.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs [hunk 5]
```diff
@@ -44,12 +43,14 @@ internal sealed partial class HttpConnection : IDisposable
         private static readonly byte[] s_spaceHttp11NewlineAsciiBytes = Encoding.ASCII.GetBytes(" HTTP/1.1\r\n");
         private static readonly byte[] s_hostKeyAndSeparator = Encoding.ASCII.GetBytes(HttpKnownHeaderNames.Host + ": ");
         private static readonly byte[] s_httpSchemeAndDelimiter = Encoding.ASCII.GetBytes(Uri.UriSchemeHttp + Uri.SchemeDelimiter);
+        private static readonly string s_cancellationMessage = new OperationCanceledException().Message;
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: Should this be in the resource?

I'm not sure what will get returned here by default.  Probably a comment is warranted.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs [hunk 6]
```diff
@@ -446,54 +462,103 @@ private Task WriteFormattedInt32Async(int value, CancellationToken cancellationT
...
+                var weakThisRef = (WeakReference<HttpConnection>)s;
+                if (weakThisRef.TryGetTarget(out HttpConnection strongThisRef))
+                {
+                    if (NetEventSource.IsEnabled) strongThisRef.Trace("Cancellation requested. Disposing of the connection.");
+                    strongThisRef.Dispose();
+                }
+            }, _weakThisRef);
+        }
+
+        private static bool ShouldWrapInOperationCanceledException(Exception error, CancellationToken cancellationToken) =>
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: This seems like a good reason to change Sockets so that they consistently throw OperationCancelledException on Close.  That seems like it would save us a fair amount of code here, right?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs [hunk 7]
```diff
@@ -1138,7 +1203,6 @@ private void ReturnConnectionToPoolCore()
                 try
                 {
                     // Null out the associated request before the connection is potentially reused by another.
-                    _currentRequest = null;
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: Why was this removed?  If it's intentional, can we update the comment?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionContent.cs [hunk 8]
```diff
@@ -11,16 +11,10 @@ namespace System.Net.Http
 {
     internal sealed partial class HttpConnection : IDisposable
     {
-        private sealed class HttpConnectionContent : HttpContent
+        private sealed class HttpConnectionResponseContent : HttpContent
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionContent.cs]: We should change filename too

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 6 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: I'm not sure how we do it consistently.  If you Dispose of the socket and then call a method on it, getting an OperationCanceledException rather than an ObjectDisposedException seems... odd.  It'll also require more try/catch/throws as we'd need to translate ObjectDisposedExceptions from SafeHandles into OperationCanceledExceptions.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionContent.cs (hunk 8 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnectionContent.cs]: Oops, yes, I meant to do that.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 7 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: Because we already did it:
https://github.com/stephentoub/corefx/blob/a137bbf42598202f999ead31d92754613145c849/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L1149

I'll remove the stray comment.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 5 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: I can copy the string to a resource.  Figured this was a straightforward way to just reuse the original message, but not a problem.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 6 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: Yeah, that makes sense.  So we will always have to translate ObjectDisposedException to OperationCancelledException.

What we could (and probably should) do, I think, is change Socket behavior to be consistent about throwing OperationCancelled instead of returning 0 bytes.  This would simplify things a bit here, right?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RawConnectionStream.cs [hunk 9]
```diff
@@ -65,14 +117,63 @@ public override Task WriteAsync(byte[] buffer, int offset, int count, Cancellati
...
+                Task flushTask = _connection.FlushAsync();
+                return flushTask.IsCompleted ?
+                    flushTask :
+                    WaitWithConnectionCancellationAsync(flushTask, cancellationToken);
+            }
 
-            public override Task FlushAsync(CancellationToken cancellationToken) =>
-                _connection != null ? _connection.FlushAsync(cancellationToken) :
-                Task.CompletedTask;
+            private async Task WaitWithConnectionCancellationAsync(Task task, CancellationToken cancellationToken)
```
Comment (geoffkizer) [inline src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RawConnectionStream.cs]: It seems like this could be used in other places as well, no?  Is the issue just avoiding the extra overhead for another async function?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 5 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: I don't care either way, just wasn't sure what was going on here.  A comment would be fine.

File: src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs [hunk 10]
```diff
@@ -2,161 +2,442 @@
...
 using System.Threading;
 using System.Threading.Tasks;
-
 using Xunit;
-using Xunit.Abstractions;
 
 namespace System.Net.Http.Functional.Tests
 {
-    public class CancellationTest : HttpClientTestBase
+    public class HttpClientHandler_CancellationTest : HttpClientTestBase
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs]: Filename?  (I don't particularly care either way, just trying to understand what the pattern is for tests generally)

File: src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs [hunk 11]
```diff
@@ -2,161 +2,442 @@
...
     {
-        private readonly ITestOutputHelper _output;
+        [OuterLoop]
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs]: Since we're using LoopbackServer, should this be inner loop?

File: src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs [hunk 12]
```diff
@@ -2,161 +2,442 @@
...
+                    await ValidateClientCancellationAsync(async () =>
+                    {
+                        var req = new HttpRequestMessage(HttpMethod.Post, url);
+                        req.Content = new DelayedByteContent(2000, 3000, preContentSent, sendPostContent.Task);
+                        req.Headers.TransferEncodingChunked = chunkedTransfer;
+                        req.Headers.ConnectionClose = connectionClose;
+
+                        Task<HttpResponseMessage> postResponse = client.SendAsync(req, HttpCompletionOption.ResponseHeadersRead, cts.Token);
+                        await preContentSent.Task;
+                        if ((mode & CancellationMode.Token) != 0) cts.Cancel();
```
Comment (geoffkizer) [inline src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs]: Nit: since you do this CancellationMode "switch" several places, you could factor it into a helper function to ensure consistency and simplify the code a bit.

**Comment by geoffkizer** [comment]
Overall, looks great!  A couple minor issues above.

Thanks for the comprehensive tests here.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs (hunk 2 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs]: If the registration cost proves troublesome, we could potentially add a "try" fast path, where we try to parse/read from what's in the buffer, and then fall back to doing all of this if we were unsuccessful.  I'm hopeful that with https://github.com/dotnet/coreclr/pull/12819 the overheads here shouldn't be too impactful.  And the costs are also largely nops if the cancellation token is None / the default; unfortunately that's never the case when going through HttpClient, but going through HttpMessageInvoker it can be.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 6 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: > What we could (and probably should) do, I think, is change Socket behavior to be consistent about throwing OperationCancelled instead of returning 0 bytes. This would simplify things a bit here, right?

Does the socket know the difference to be able to make that call?

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RawConnectionStream.cs (hunk 9 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/RawConnectionStream.cs]: > Is the issue just avoiding the extra overhead for another async function?

Right.  If we had something like https://github.com/dotnet/roslyn/issues/15491, then we would definitely reuse it in more places.

File: src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs (hunk 10 above)
Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs]: > just trying to understand what the pattern is for tests generally

We're all over the place.  I can change the file name.

File: src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs (hunk 11 above)
Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs]: > But historically, cancelation tests tend to be unreliable in general because they rely on timing.

These shouldn't be unreliable, as they use synchronization between client and server to order operations as desired.  There are some places where cancellation may occur here or there and thus we're not necessarily testing exactly the thing we hope to test, but the tests should still pass and be reliable from that perspective.

> Since we're using LoopbackServer, should this be inner loop?

I'll need to look again at how long the various tests take.  There are reasons besides loopback server to make things outerloop, e.g. they're not super fast, but IIRC all of these tests run within a second or two, so innerloop should be ok.

File: src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs (hunk 12 above)
Comment (stephentoub) [reply src/System.Net.Http/tests/FunctionalTests/CancellationTest.cs]: Yup

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs (hunk 2 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs]: Ok, seems reasonable to me.

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs (hunk 6 above)
Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: I think it should be able to.  We can do what we're doing here: track that we're closing, and if the operation returns EOF, we just treat it as a cancelled op.

Changing Sockets to be consistent would be good regardless, if we can do it.

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: But we're not passing the cancellation token to the socket, so it doesn't know if cancellation was requested.

Comment (geoffkizer) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: The latter, yeah.

Maybe I'm remembering wrong, but I seem to recall that the issue here is that Linux behaves different than Windows: Windows throws an error, Linux just returns EOF.  So the goal here would be to get consistent, and also to simplify HttpClient usage here.

I could be wrong though... you've been testing this scenario, what have you found?

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: > you've been testing this scenario, what have you found?

I believe that lines up with what I saw... I had to add the EOF handling once I started testing on Linux (though I can't say for sure I never saw it on Windows).

Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs]: https://github.com/dotnet/corefx/issues/27066

File: src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs (hunk 2 above)
Comment (stephentoub) [reply src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/ChunkedEncodingReadStream.cs]: One option I thought of is to use a custom awaiter, e.g.
```C#
private struct WithConfigureAwaitAndCancellation<T> : ICriticalNotifyCompletion
{
    private readonly ConfiguredValueTaskAwaitable<T>.ConfiguredValueTaskAwaiter _awaiter;
    private readonly CancellationToken _token;
    private readonly HttpConnection _connection;
    private CancellationTokenRegistration _ctr;

    public WithConfigureAwaitAndCancellation(
        ConfiguredValueTaskAwaitable<T>.ConfiguredValueTaskAwaiter awaiter,
        HttpConnection connection,
        CancellationToken cancellationToken)
    {
        _awaiter = awaiter;
        _connection = connection;
        _token = cancellationToken;
        _ctr = default;
    }

    public WithConfigureAwaitAndCancellation<T> GetAwaiter() => this;

    public void UnsafeOnCompleted(Action continuation)
    {
        _ctr = _connection.RegisterCancellation(_token);
        _awaiter.UnsafeOnCompleted(continuation);
    }

    public void OnCompleted(Action continuation) => UnsafeOnCompleted(continuation);

    public bool IsCompleted => _awaiter.IsCompleted;

    public T GetResult()
    {
        _ctr.Dispose();
        return _awaiter.GetResult();
    }
}
```
and use that when awaiting on each of the operations.  That way, we only end up registering with the token if we actually yield (if IsCompleted returns false).  I'm not going to do that now, but we can try it out subsequently to see if it makes a difference.

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Linux x64 Release Build please

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Linux x64 Release Build please

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build please

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Linux x64 Release Build please

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Linux x64 Release Build please

**Comment by stephentoub** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build please
@dotnet-bot test Outerloop NETFX x86 Release Build please

**Comment by stephentoub** [comment]
The Windows outerloop failure is a timeout we've seen before a lot on Win7.  I've run outerloop tests now several times on Linux, Windows, and netfx, and the timeouts I thought I was seeing before due to this change haven't reproed, nor have I been able to repro locally.  Hopefully they were transient.
